### ISSUE TITLE
Segmentation fault (core dumped) 

### ISSUE THREAD

**Issue by ShenBanigans** [zmap/zmap#702]
# Describe the bug

ZMAP3 will periodically, and randomly segfault. The command-line options used do not seem to mater. **This would also happen with ZMAP2 which is why I have started trying out ZMAP3 but it still happens in version 3**. It is extremely difficult for me to provide information to reproduce the issue due the way in which I am utilizing your tool for research but I will do my best: 

I have a built a python script for scanning large numbers of subnets for specific ports, it organizes the folder structure the way I want and reads a list of subnets in CIDR format from a specified text file and loops through a list of ports I want to scan and then uses python subprocess module to execute zmap against the subnet for the port I want. Say I have 2700 subnets in the input file, it will sometimes segfault after getting through a couple of hundred, sometimes it cant even get through 10 invocations, sometimes it will run for hours with no problems. It is an extremely bug since I cannot automate my research as I have to constantly check on if zmap has crashed or not when invoked by python.

I have used zmap for decades and have only started noticing this bug during a research task that involves scanning thousands of subnets. **Please note that it does not matter what command line arguments I am using with ZMAP and it also does not matter what subnets are being scanned** - For example say I have a list of 2500 CIDR ranges and I am looping through the list and say for example it segfaults on the 10th one in the list, if i were to re-run my script which loops through the list and invokes zmap, it wont segfault again at the 10th one in the list, it might way past that before it happens again, it might segfault on the first one in the list, it might ssegfault on the 200th one etc.

# CLI Arguments
I am invoking zmap using a python script with subprocess.check_call() and the command line arguments I use vary depending on what I am wanting to do with a specific subnet. However here is an example of a segfault situation where it crashed out when scanning the 4th subnet out of a list of 1024 total. **Please note the bug happens when scanning TCP or UDP, on any port, for any subnet size, the below is just an example of one time it happened when I was scanning for UDP/161.** 

## Example One

```
zmap -M udp --probe-args=file:/home/user/Downloads/zmap-3.0.0-beta1/examples/udp-probes/snmp1_161.pkt -i enp0s31f6 -T 1 -B 1M -p 161 --output-filter="success = 1" -O csv 41.72.64.0/19 -o 41.72.64.0_19-161
Oct 13 14:22:48.767 [WARN] blocklist: ZMap is currently using the default blocklist located at /etc/zmap/blocklist.conf. By default, this blocklist excludes locally scoped networks (e.g. 10.0.0.0/8, 127.0.0.1/8, and 192.168.0.0/16). If you are trying to scan local networks, you can change the default blocklist by editing the default ZMap configuration at /etc/zmap/zmap.conf.
 0:00 0%; send: 4 1 p/s (1.07 Kp/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:01 9%; send: 1171 1.17 Kp/s (1.17 Kp/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:02 17%; send: 2313 1.14 Kp/s (1.15 Kp/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:03 26%; send: 3455 1.14 Kp/s (1.15 Kp/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:04 34%; send: 4096 done (1.11 Kp/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:05 43% (7s left); send: 4096 done (1.11 Kp/s avg); recv: 1 1 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.02%
 0:06 51% (6s left); send: 4096 done (1.11 Kp/s avg); recv: 1 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.02%
 0:07 60% (5s left); send: 4096 done (1.11 Kp/s avg); recv: 1 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.02%
Segmentation fault (core dumped)

**dmesg output:** [1201962.947108] zmap[2426399]: segfault at 6348658f ip 000055d2727b04e5 sp 00007ff7e14079b0 error 4 in zmap[55d27279e000+1d000]
```
## Example Two

```
zmap -i enp0s31f6 -T 1 -B 1M -p 8080 --output-filter="success = 1" -O csv 62.249.153.0/25 -o 62.249.153.0_25-8080
Oct 13 14:22:48.767 [WARN] blocklist: ZMap is currently using the default blocklist located at /etc/zmap/blocklist.conf. By default, this blocklist excludes locally scoped networks (e.g. 10.0.0.0/8, 127.0.0.1/8, and 192.168.0.0/16). If you are trying to scan local networks, you can change the default blocklist by editing the default ZMap configuration at /etc/zmap/zmap.conf.
 0:00 0%; send: 4 1 p/s (1.07 Kp/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
 0:01 9%; send: 1171 1.17 Kp/s (1.17 Kp/s avg); recv: 0 0 p/s (0 p/s avg); drops: 0 p/s (0 p/s avg); hitrate: 0.00%
Segmentation fault (core dumped)
```

# Example Target IP

As previously explained I have a text file containing a list of CIDR subnets. This next file can conatin anywhere between 2500 to 60,000 subnets. I know that the issue does not seem to be related to the target subnet. I do not want to paste such an enormous list here, however if you would like to try to reproduce this pick a country from here [https://www.nirsoft.net/countryip/](url) and convert those ranges to CIDR and then loop through them, invoking ZMAP as appropriate.

# Expended behavior
Not segfault.

## Things I have tried to remediate the issue (none worked)

- Specifying the speed using -r for rate instead of -B for bandwidth and trying all kinds of bandwidth and pps rates.
- Using different amount of threads.
- Pining to different cores.
- Scanning using WiFi and Ethernet interfaces on the machine invoking ZMAP.
- Scanning within a virtual machine.
- Scanning using Ubuntu, Debian, Denuvo and Arch Linux (trying differen distributions)
- Using a completely different physical machine.
- Looping through my text file of targets with bash instead of python so that it is invoked differently. 

# Environment

- As stated above I have tried numerous different distributions, however the two examples given above were from **Ubuntu 22.04.1 LTS**
 - Version: Built from source [](https://github.com/zmap/zmap/releases/tag/v3.0.0-beta1) (this is the only source I can find for zmap3, please correct me if I am wrong).

# Other 

- Is ZMAP 3 still maintained or is this dead now? I hope not since after years of using many similar tools zmap is the most accurate for the specific activities I do.

**Comment by ShenBanigans** [comment]
Sorry, my bad. This was due to calling zmap too quickly, and it would core dump when it cannot open a socket on the system.

**Comment by zakird** [comment]
Any chance that you have a backtrace or some information about how you came to that conclusion? This is certainly still a bug we should try to fix.

**Comment by dadrian** [comment]
I believe this was resolved in #705
